File: python/src/semantic_retrieval/transformation/document/text/text_chunk_transformer.py
from hashlib import md5
from typing import List, Optional
from uuid import uuid4

from semantic_retrieval.common.types import CallbackEvent, Record
from semantic_retrieval.document.document import (
    Document,
    DocumentFragmentType,
    TransformedDocument,
)
from semantic_retrieval.document.metadata.document_metadata_db import (
    DocumentMetadataDB,
)
from semantic_retrieval.transformation.document.document_transformer import (
    BaseDocumentTransformer,
)
from semantic_retrieval.utils.callbacks import CallbackManager, Traceable


class TextChunkConfig(Record):
    chunk_size_limit: int
    chunk_overlap: int


class TextChunkTransformerParams:
    separator: str = " "
    strip_new_lines: bool = True
    metadata_db: Optional[DocumentMetadataDB]


async def _len(x: str) -> int:
    return len(x)


class TextChunkTransformer(BaseDocumentTransformer, Traceable):
    def __init__(self, callback_manager: CallbackManager):
        self.size_fn = _len
        self.chunk_size_limit = 500
        self.chunk_overlap = 100
        self.callback_manager = callback_manager

    async def chunk_text(self, text: str) -> List[str]:
        raise NotImplementedError(
            "This method must be implemented in a derived class"
        )

    def sub_chunk_on_separator(self, text: str, separator: str) -> List[str]:
        sub_chunks = text.split(separator)
        return [sc for sc in sub_chunks if sc != ""]

    async def transform_document(self, document: Document) -> Document:
        original_fragments_data = [
            {
                "attributes": fragment.attributes,
                "content": await fragment.get_content(),
                "fragmentType": fragment.fragment_type,
                "metadata": fragment.metadata,
            }
            for fragment in document.fragments
        ]

        transformed_fragments = []
        fragment_count = 0
        document_id = str(uuid4())

        for original_fragment_data in original_fragments_data:
            original_fragment = original_fragment_data["content"]

            def id_(chunk: str) -> str:
                return chunk

            # TODO [P1]: One other issue with chunks is some csvs are getting 0 chunks returned even though they have sub_chunks
            # Definitely some issue with merge_sub_chunks, but not going to debug this right now
            if not isinstance(original_fragment, str):
                raise ValueError(
                    f"Expected original_fragment to be a string, but got {type(original_fragment)}"
                )
            for chunk in await self.chunk_text(original_fragment):
                current_fragment = {
                    "fragment_id": str(uuid4()),
                    "fragment_type": DocumentFragmentType.TEXT,
                    "document_id": document_id,
                    "metadata": original_fragment_data["metadata"],
                    "attributes": {},
                    "hash": md5(chunk.encode()).hexdigest(),
                    "content": chunk,
                    "getContent": id_,
                    "serialize": id_,
                }

                # TODO [P1]: Unsure if this is working correctly
                if fragment_count > 0:
                    transformed_fragments[fragment_count - 1][
                        "nextFragment"
                    ] = current_fragment

                fragment_count += 1
                transformed_fragments.append(current_fragment)

        transformed_document = TransformedDocument(
            document=document,
            document_id=document_id,
            fragments=transformed_fragments,
            collection_id=document.collection_id,
            metadata={},
            attributes={},
        )

        await self.callback_manager.run_callbacks(
            CallbackEvent(
                name="text_chunk_transform_document",
                data=dict(
                    original_document=document,
                    transformed_document=transformed_document,
                    original_fragments_data=original_fragments_data,
                ),
            )
        )

        return transformed_document

    def join_sub_chunks(
        self, sub_chunks: List[str], separator: str
    ) -> Optional[str]:
        chunk = separator.join(sub_chunks).strip()
        return chunk if chunk != "" else None

    async def merge_sub_chunks(
        self, sub_chunks: List[str], separator: str
    ) -> List[str]:
        chunks = []
        prev_sub_chunks = []
        current_sub_chunks = []
        current_chunk_size = 0
        chunk_separator_size = await self.size_fn(separator)

        for sub_chunk in sub_chunks:
            sub_chunk_size = await self.size_fn(sub_chunk)

            if sub_chunk_size > self.chunk_size_limit:
                print(
                    f"SubChunk size {sub_chunk_size} exceeds chunkSizeLimit of {self.chunk_size_limit}"
                )

            if (
                current_chunk_size + chunk_separator_size + sub_chunk_size
                > self.chunk_size_limit
            ):
                chunk = self.join_sub_chunks(current_sub_chunks, separator)
                if chunk is not None:
                    chunks.append(chunk)
                prev_sub_chunks = current_sub_chunks
                current_chunk_size = 0
                current_sub_chunks = []

            num_total_prev_sub_chunks = len(prev_sub_chunks)

            if len(current_sub_chunks) == 0 and num_total_prev_sub_chunks > 0:
                prev_sub_chunks_overlap_size = 0
                num_prev_sub_chunks_overlap = 0

                while num_prev_sub_chunks_overlap < num_total_prev_sub_chunks:
                    next_prev_sub_chunk_size = await self.size_fn(
                        prev_sub_chunks[
                            num_total_prev_sub_chunks
                            - num_prev_sub_chunks_overlap
                            - 1
                        ]
                    )

                    if (
                        prev_sub_chunks_overlap_size
                        + chunk_separator_size
                        + next_prev_sub_chunk_size
                        > self.chunk_overlap
                    ) or (
                        prev_sub_chunks_overlap_size
                        + chunk_separator_size
                        + next_prev_sub_chunk_size
                        + sub_chunk_size
                        > self.chunk_size_limit
                    ):
                        break

                    prev_sub_chunks_overlap_size += next_prev_sub_chunk_size

                    if num_prev_sub_chunks_overlap > 0:
                        prev_sub_chunks_overlap_size += chunk_separator_size

                    num_prev_sub_chunks_overlap += 1

                while num_prev_sub_chunks_overlap > 0:
                    current_sub_chunks.append(
                        prev_sub_chunks[
                            num_total_prev_sub_chunks
                            - num_prev_sub_chunks_overlap
                        ]
                    )
                    num_prev_sub_chunks_overlap -= 1

                current_chunk_size = prev_sub_chunks_overlap_size

            current_sub_chunks.append(sub_chunk)

            if len(current_sub_chunks) > 1:
                current_chunk_size += chunk_separator_size

            current_chunk_size += sub_chunk_size

        if len(current_sub_chunks) > 0:
            chunk = self.join_sub_chunks(current_sub_chunks, separator)

        await self.callback_manager.run_callbacks(
            CallbackEvent(
                name="merge_sub_chunks",
                data=dict(
                    sub_chunks=sub_chunks,
                    separator=separator,
                    chunks=chunks,
                ),
            )
        )

        # TODO [P1] is this correct?
        return chunks

File: typescript/src/transformation/document/text/textChunkTransformer.ts
import { BaseDocumentTransformer } from "../documentTransformer";
import { Document, DocumentFragment } from "../../../document/document";
import { DocumentMetadataDB } from "../../../document/metadata/documentMetadataDB";
import { v4 as uuid } from "uuid";
import { Md5 } from "ts-md5";
import { Traceable, TransformDocumentEvent } from "../../../utils/callbacks";

export interface TextChunkConfig {
  // The size limit of each chunk, measured by sizeFn. Each chunk (including overlap)
  // should not exceed this limit, but may be smaller
  chunkSizeLimit?: number;
  // When constructing a new chunk from sub-chunks and the first sub-chunk is small enough,
  // the new chunk will have the latest sub-chunks from the previous chunk prepended to it until
  // chunkOverlap is reached. This allows for more context to be included in each chunk.
  chunkOverlap?: number;
  // Determines how size is measured for the text. Defaults to text length.
  sizeFn?: (text: string) => Promise<number>;
}

export type TextChunkTransformerParams = TextChunkConfig &
  Traceable & {
    metadataDB?: DocumentMetadataDB;
  };

/**
 * A DocumentTransformer that splits a Document's fragments into chunks of
 * text of a maximum size (chunkSize). The chunking operation consists of a
 * series of splits (chunk larger text into smaller sub-chunks) and merges (merge
 * smaller sub-chunks into one larger chunk).
 */
export abstract class TextChunkTransformer
  extends BaseDocumentTransformer
  implements TextChunkConfig
{
  chunkSizeLimit = 500;
  chunkOverlap = 100;
  sizeFn = async (text: string) => text.length;

  constructor(params?: TextChunkTransformerParams) {
    super(params?.metadataDB, params?.callbackManager);
    this.chunkSizeLimit = params?.chunkSizeLimit ?? this.chunkSizeLimit;
    this.chunkOverlap = params?.chunkOverlap ?? this.chunkOverlap;
    this.sizeFn = params?.sizeFn ?? this.sizeFn;

    if (this.chunkOverlap >= this.chunkSizeLimit) {
      throw new Error("chunkOverlap must be less than chunkSizeLimit");
    }
  }

  abstract chunkText(text: string): Promise<string[]>;

  protected subChunkOnSeparator(text: string, separator: string): string[] {
    const subChunks = text.split(separator);
    return subChunks.filter((sc) => sc !== "");
  }

  async transformDocument(document: Document): Promise<Document> {
    const originalFragmentsData = await Promise.all(
      document.fragments.map(async (fragment) => {
        const content = await fragment.getContent();
        return {
          attributes: fragment.attributes,
          content,
          fragmentType: fragment.fragmentType,
          metadata: fragment.metadata,
        };
      })
    );

    const transformedFragments: DocumentFragment[] = [];
    let fragmentCount = 0;
    const documentId = uuid();

    for (let i = 0; i < originalFragmentsData.length; i++) {
      const originalFragment = originalFragmentsData[i];

      for (const chunk of await this.chunkText(originalFragment.content)) {
        const currentFragment: DocumentFragment = {
          fragmentId: uuid(),
          fragmentType: "text",
          documentId,
          // TODO: Figure out if/how we should add additional metadata from the chunks (e.g. line count)
          metadata: originalFragment.metadata,
          attributes: {},
          // TODO: figure out blobId
          hash: Md5.hashStr(chunk),
          getContent: async () => chunk,
          serialize: async () => chunk,
          previousFragment:
            fragmentCount > 0
              ? transformedFragments[fragmentCount - 1]
              : undefined,
        };

        if (fragmentCount > 0) {
          transformedFragments[fragmentCount - 1].nextFragment =
            currentFragment;
        }

        fragmentCount++;
        transformedFragments.push(currentFragment);
      }
    }

    const transformedDocument = {
      ...document,
      documentId,
      fragments: transformedFragments,
    };

    const event: TransformDocumentEvent = {
      name: "onTransformDocument",
      originalDocument: document,
      transformedDocument,
    };
    await this.callbackManager?.runCallbacks(event);

    // TODO: Think through metadata handling, since setting new doc metadata on each transformation
    // can cause proliferation of DB entries. On the other hand, we probably don't want to mutate
    // the document in place since we may want to perform different operations on the same document.
    // With optional metadata DB, we can just pass it through in the 'final' transformations / those
    // we want to persist document metadata for. But, access policies are currently constructed from the raw document
    // during parsing, so we'd need to support the same policy factory stuff in each transformer...
    return transformedDocument;
  }

  protected joinSubChunks(
    subChunks: string[],
    separator: string
  ): string | null {
    const chunk = subChunks.join(separator).trim();
    return chunk === "" ? null : chunk;
  }

  protected async mergeSubChunks(
    subChunks: string[],
    separator: string
  ): Promise<string[]> {
    const chunks: string[] = [];
    let prevSubChunks: string[] = [];
    let currentSubChunks: string[] = [];
    let currentChunkSize = 0;

    const chunkSeparatorSize = await this.sizeFn(separator);

    for (const sc of subChunks) {
      const scSize = await this.sizeFn(sc);

      if (scSize > this.chunkSizeLimit) {
        console.warn(
          `SubChunk size ${scSize} exceeds chunkSizeLimit of ${this.chunkSizeLimit}`
        );
      }

      if (
        currentChunkSize + chunkSeparatorSize + scSize >
        this.chunkSizeLimit
      ) {
        // Current subChunk would cause the chunk (with separators) to exceed the size limit, so
        // create the current chunk before starting a new one with the current subChunk
        const chunk = this.joinSubChunks(currentSubChunks, separator);
        if (chunk != null) {
          chunks.push(chunk);
        }
        prevSubChunks = currentSubChunks;
        currentChunkSize = 0;
        currentSubChunks = [];
      }

      // If this is the first subChunk after a completed chunk, handle chunk overlap from prev chunk
      const numTotalPrevSubChunks = prevSubChunks.length;
      if (currentSubChunks.length === 0 && numTotalPrevSubChunks > 0) {
        let prevSubChunksOverlapSize = 0;
        let numPrevSubChunksOverlap = 0;

        // We take as many previous subchunks that will fit into the overlap along with at least one current
        // chunk, as long as they all fit within the chunk size limit
        while (numPrevSubChunksOverlap < numTotalPrevSubChunks) {
          const nextPrevSubChunkSize = await this.sizeFn(
            prevSubChunks[numTotalPrevSubChunks - numPrevSubChunksOverlap - 1]
          );

          if (
            prevSubChunksOverlapSize +
              chunkSeparatorSize +
              nextPrevSubChunkSize >
              this.chunkOverlap ||
            prevSubChunksOverlapSize +
              chunkSeparatorSize +
              nextPrevSubChunkSize +
              scSize >
              this.chunkSizeLimit
          ) {
            // Adding this prev subchunk would exceed the overlap or chunk size limit, so stop
            break;
          }

          prevSubChunksOverlapSize += nextPrevSubChunkSize;
          if (numPrevSubChunksOverlap > 0) {
            // Separator only added between subchunks, so skip for first subchunk
            prevSubChunksOverlapSize += chunkSeparatorSize;
          }
          numPrevSubChunksOverlap++;
        }

        while (numPrevSubChunksOverlap > 0) {
          currentSubChunks.push(
            prevSubChunks[numTotalPrevSubChunks - numPrevSubChunksOverlap]
          );
          numPrevSubChunksOverlap--;
        }

        currentChunkSize = prevSubChunksOverlapSize;
      }

      currentSubChunks.push(sc);
      if (currentSubChunks.length > 1) {
        // Separator only added between subchunks, so skip for first subchunk
        currentChunkSize += chunkSeparatorSize;
      }
      currentChunkSize += scSize;
    }

    // Close up the final chunk if needed
    if (currentSubChunks.length > 0) {
      const chunk = this.joinSubChunks(currentSubChunks, separator);
      if (chunk != null) {
        chunks.push(chunk);
      }
    }

    return chunks;
  }
}
